edb/server/pgcon/rust_transport.py

```diff
@@ -0,0 +1,475 @@
+    sslmode = SSLMode.parse(ssl_config['sslmode'])
+    ssl = ssl_module.SSLContext(ssl_module.PROTOCOL_TLS_CLIENT)
+    ssl.check_hostname = sslmode >= SSLMode.verify_full
+    if sslmode < SSLMode.require:
+        ssl.verify_mode = ssl_module.CERT_NONE
+    else:
+        if ssl_config['sslrootcert']:
+            ssl.load_verify_locations(ssl_config['sslrootcert'])
+            ssl.verify_mode = ssl_module.CERT_REQUIRED
+        else:
+            if sslmode == SSLMode.require:
+                ssl.verify_mode = ssl_module.CERT_NONE
+        if ssl_config['sslcrl']:
+            ssl.load_verify_locations(ssl_config['sslcrl'])
+            ssl.verify_flags |= ssl_module.VERIFY_CRL_CHECK_CHAIN
+    if ssl_config['sslkey'] and ssl_config['sslcert']:
+        ssl.load_cert_chain(ssl_config['sslcert'],
+                            ssl_config['sslkey'],
+                            ssl_config['sslpassword'] or '')
+    if ssl_config['ssl_max_protocol_version']:
+        ssl.maximum_version = _parse_tls_version(
+            ssl_config['ssl_max_protocol_version'])
+    if ssl_config['ssl_min_protocol_version']:
+        ssl.minimum_version = _parse_tls_version(
+            ssl_config['ssl_min_protocol_version'])
+    # OpenSSL 1.1.1 keylog file
+    if hasattr(ssl, 'keylog_filename'):
+        if ssl_config['keylog_filename']:
+            ssl.keylog_filename = ssl_config['keylog_filename']
+    return ssl
+
+
+class PGConnectionProtocol(asyncio.Protocol):
+    """A protocol that manages the initial connection and authentication process
+    for PostgreSQL.
+
+    This protocol acts as an intermediary between the raw socket connection and
+    the user's protocol.
+    """
+    def __init__(
+            self,
+            state: PyConnectionState,
+            protocol: asyncio.Protocol,
+            ready_future: asyncio.Future,
+            pg_state: PGState,
+        ):
+        self.state = state
+        self.pg_state = pg_state
+        self.protocol = protocol
+        self.ready_future = ready_future
+        self._writing_paused = False
+        self._ready = False
+
+    def data_received(self, data: bytes):
+        if self._ready:
+            self.protocol.data_received(data)
+        else:
+            try:
+                self.state.drive_message(memoryview(data))
+            except Exception as e:
+                if error := self.pg_state.server_error:
+                    self.pg_state.server_error = None
+                    self.ready_future.set_exception(pgerror.BackendConnectionError(fields=dict(error)))
+                else:
+                    self.ready_future.set_exception(ConnectionError(e))
+            if self.state.is_ready():
+                self._ready = True
+                self.ready_future.set_result(True)
+
+    def connection_lost(self, exc):
+        if self._ready:
+            self.protocol.connection_lost(exc)
+        else:
+            if not self.ready_future.done:
```

Comment:
`done` is a method - it'll be always true if not called:

```suggestion
            if not self.ready_future.done():
```